### PR TITLE
Optional availability calculation mode to allow planned maintenance

### PR DESCRIPTION
**Pull request by ottorei**
Please give a short description what your pull request is for

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

Adds an optional availability calculation mode where scheduled maintenance does not affect availability. The option can be enabled from options and is disabled by default.

#### Screenshots

optional-outage-mode

#### Additional changes

Do not use device's uptime anymore while calculating availability. If the device is not reachable to the monitoring system, consider it unavailable.

#### WIP status

Needs more testing. Tested with the following cases:

Device not in maintenance mode

1. Device goes down => an outage should start
2. Device come back up => the latest outage ends

Optional mode disabled => PASS
Optional mode enabled => PASS

Device in maintenance mode, outage is shorter than maintenance break

1. Device goes down => no outage should start
2. Device comes back up => no modifications should occur to the outages table

Optional mode disabled => PASS
Optional mode enabled => PASS

Device in maintenance mode but outage lasts longer than maintenance break

1. Device goes down => no outage yet
2. Time passes until maintenance break is over => an outage should start
3. Device comes back up => outage ends

Optional mode disabled => PASS
Optional mode enabled => PASS

### CODE REVIEW

**Comment by ottorei** [comment]
Some test cases:

# Device not in maintenance mode (optional mode enabled)
1. Device goes down => an outage should start with current timestamp, an alert is issued
2. Device come back up => the latest outage ends

Optional mode enabled
==> Pass

Optional mode disabled
==> Pass

-----------------------------------------------

# Device in maintenance mode, outage is shorter than maintenance break (optional mode enabled)
1. Device goes down => no outage should start
2. Device comes back up => no modifications should occur to the outages table and no alerts should be issued
3. Alerts are not issued

Optional mode enabled
==> Pass

Optional mode disabled
==> Pass

-----------------------------------------------

# Device in maintenance mode but outage lasts longer than maintenance break
1. Device goes down => no outage yet
2. Time passes until maintenance break is over => an outage should start with current timestamp and an alert is issued
3. Device comes back up => outage ends

Optional mode enabled
==> Pass

Optional mode disabled
==> Pass

**Comment by SourceDoctor** [comment]
could you fix CI Errors please?

File: includes/functions.php [hunk 1]
```diff
@@ -2000,6 +2000,8 @@ function device_is_up($device, $record_perf = false)
     $device_perf = $ping_response['db'];
     $device_perf['device_id'] = $device['device_id'];
     $device_perf['timestamp'] = ['NOW()'];
+    $maintenance = DeviceCache::get($device['device_id'])->isUnderMaintenance();
+    $mode = Config::get('graphing.availability_ignore_maintenance');
```
Comment (SourceDoctor) [inline includes/functions.php:2004]: variable name should be more clear
maybe: "consider_maintenance mode"

File: includes/functions.php [hunk 2]
```diff
@@ -2064,13 +2080,17 @@ function device_is_up($device, $record_perf = false)
...
+        else {
             $type = 'down';
             $reason = $response['status_reason'];
 
+            // use current time as a starting point when an outage starts
             $data = ['device_id' => $device['device_id'],
-                'going_down' => strtotime($device['last_polled']), ];
-            dbInsert($data, 'device_outages');
+                'going_down' => time(), ];
+            if ( (! $maintenance && $mode) || ($maintenance && ! $mode) || (! $maintenance && ! $mode)) {
```
Comment (SourceDoctor) [inline includes/functions.php:2091]: seems more correct to me, but might be wrong:

`if ( ! $mode || (!$maintenance && $mode)) {`

`if ( ($maintenance xor $mode) || (! $maintenance && ! $mode)) {`

File: includes/functions.php [hunk 3]
```diff
@@ -2064,13 +2080,17 @@ function device_is_up($device, $record_perf = false)
...
                 );
             }
-        } else {
+        }
+        else {
             $type = 'down';
             $reason = $response['status_reason'];
 
+            // use current time as a starting point when an outage starts
             $data = ['device_id' => $device['device_id'],
```
Comment (SourceDoctor) [inline includes/functions.php:2089]: i would include this line into if clause, also

File: includes/functions.php [hunk 4]
```diff
@@ -2064,13 +2080,17 @@ function device_is_up($device, $record_perf = false)
...
-                'going_down' => strtotime($device['last_polled']), ];
-            dbInsert($data, 'device_outages');
+                'going_down' => time(), ];
```
Comment (SourceDoctor) [inline includes/functions.php:2090]: i would move this into if clause also

File: includes/functions.php [hunk 5]
```diff
@@ -2025,15 +2027,29 @@ function device_is_up($device, $record_perf = false)
...
             $response['status_reason'] = '';
-        } else {
+        }
+        else {
             echo 'SNMP Unreachable';
             $response['status'] = '0';
-            $response['status_reason'] = 'snmp';
+            if ($maintenance && $mode) {
+                // Scheduled maintenance, device not responding to snmp
+                $response['status_reason'] = 'snmp (maintenance)';
```
Comment (SourceDoctor) [inline includes/functions.php:2036]: have you rechecked the system, to avoid conflicts through this new state?

File: includes/functions.php [hunk 6]
```diff
@@ -2025,15 +2027,29 @@ function device_is_up($device, $record_perf = false)
...
         }
-    } else {
+    } 
+    else {
         echo 'Unpingable';
         $response['status'] = '0';
-        $response['status_reason'] = 'icmp';
+        if($maintenance && $mode) {
+            // Scheduled maintenance, device not responding to icmp
+            $response['status_reason'] = 'icmp (maintenance)';
```
Comment (SourceDoctor) [inline includes/functions.php:2048]: have you rechecked the system, to avoid conflicts through this new state?

File: includes/functions.php (hunk 2 above)
Comment (ottorei) [reply includes/functions.php:2091]: That should work since that also covers all three cases

File: includes/functions.php (hunk 1 above)
Comment (ottorei) [reply includes/functions.php]: Changed to availability_consider_maintenance

File: includes/functions.php (hunk 5 above)
Comment (ottorei) [reply includes/functions.php:2036]: I did some testing and could not find any issues yet

File: includes/functions.php (hunk 6 above)
Comment (ottorei) [reply includes/functions.php:2048]: I did some testing and could not find any issues yet

File: includes/functions.php (hunk 5 above)
Comment (SourceDoctor) [reply includes/functions.php:2035]: i grepped through system for status_reason and found many possible upcoming problems. A thinkover would be useful in my eyes.

Comment (ottorei) [reply includes/functions.php]: I will try another way to detect this state change without modifying the status_reason, WIP

**Review by murrant — APPROVED** [review]

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-69-release-changelog-october-2020/13837/1

**Pull request merged**